Spigot-Server-Patches/0726-Add-Unix-domain-socket-support.patch

```diff
@@ -0,0 +1,135 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Andrew Steinborn <git@steinborn.me>
+Date: Tue, 11 May 2021 17:39:22 -0400
+Subject: [PATCH] Add Unix domain socket support
+
+
+diff --git a/src/main/java/net/minecraft/network/NetworkManager.java b/src/main/java/net/minecraft/network/NetworkManager.java
+index f86f430598026a3a7e27fb8d40cfc5fe7b9b845d..897bb4af47c9cb6b6a7d89ab10e1eab0e80c71b9 100644
+--- a/src/main/java/net/minecraft/network/NetworkManager.java
++++ b/src/main/java/net/minecraft/network/NetworkManager.java
+@@ -564,6 +564,11 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+     // Spigot Start
+     public SocketAddress getRawAddress()
+     {
++        // Paper start - this can be nullable in the case of a Unix domain socket, so if it is, fake something
++        if (this.channel.remoteAddress() == null) {
++            return new java.net.InetSocketAddress(java.net.InetAddress.getLoopbackAddress(), 25565);
++        }
++        // Paper end
+         return this.channel.remoteAddress();
+     }
+     // Spigot End
+diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+index df8270c40ed7ce6f628686ff6f4fa4cf96af6738..153849995f4fd0e629568dbce58e405807b7b754 100644
+--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+@@ -225,6 +225,20 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
+         this.i(dedicatedserverproperties.enforceWhitelist);
+         // this.saveData.setGameType(dedicatedserverproperties.gamemode); // CraftBukkit - moved to world loading
+         DedicatedServer.LOGGER.info("Default game type: {}", dedicatedserverproperties.gamemode);
++        // Paper start - Unix domain socket support
++        java.net.SocketAddress bindAddress;
++        if (this.getServerIp().startsWith("unix:")) {
++            if (!io.netty.channel.epoll.Epoll.isAvailable()) {
++                DedicatedServer.LOGGER.warn("**** INVALID CONFIGURATION!");
++                DedicatedServer.LOGGER.warn("You are trying to use a Unix domain socket but you're not on a supported OS.");
++                return false;
++            } else if (!com.destroystokyo.paper.PaperConfig.velocitySupport && !org.spigotmc.SpigotConfig.bungee) {
++                DedicatedServer.LOGGER.warn("**** INVALID CONFIGURATION!");
++                DedicatedServer.LOGGER.warn("Unix domain sockets require IPs to be forwarded from a proxy such as BungeeCord or Velocity.");
```

Comment:
Perhaps mention Velocity first, because AFAIK BungeeCord doesn't support Unix sockets?

Reply:
I'll just omit mention of any specific proxy instead. BungeeCord does actually support listening on a Unix domain socket, but it connects to servers over TCP/IP.